compiler/generator_dart.c

```diff
@@ -0,0 +1,1274 @@
+            write_int(g, g->failure_label);
+            write_string(g, "();");
+            g->unreachable = true;
+    }
+    write_newline(g);
+}
+
+static void write_failure_if(struct generator * g, char * s, struct node * p) {
+
+    writef(g, "~Mif (", p);
+    writef(g, s, p);
+    writef(g, ")~N", p);
+    write_block_start(g);
+    write_failure(g);
+    write_block_end(g);
+    g->unreachable = false;
+}
+
+/* if at limit fail */
+static void write_check_limit(struct generator * g, struct node * p) {
+
+    if (p->mode == m_forward) {
+        write_failure_if(g, "cursor >= limit", p);
+    } else {
+        write_failure_if(g, "cursor <= limit_backward", p);
+    }
+}
+
+/* Formatted write. */
+static void writef(struct generator * g, const char * input, struct node * p) {
+    int i = 0;
+    int l = strlen(input);
+
+    while (i < l) {
+        int ch = input[i++];
+        if (ch != '~') {
+            write_char(g, ch);
+            continue;
+        }
+        switch (input[i++]) {
+            default: write_char(g, input[i - 1]); continue;
+            case 'C': write_comment(g, p); continue;
+            case 'f': write_block_start(g);
+                      write_failure(g);
+                      g->unreachable = false;
+                      write_block_end(g);
+                      continue;
+            case 'M': write_margin(g); continue;
+            case 'N': write_newline(g); continue;
+            case '{': write_block_start(g); continue;
+            case '}': write_block_end(g); continue;
+            case 'S': write_string(g, g->S[input[i++] - '0']); continue;
+            case 'B': write_b(g, g->B[input[i++] - '0']); continue;
+            case 'I': write_int(g, g->I[input[i++] - '0']); continue;
+            case 'V': write_varref(g, g->V[input[i++] - '0']); continue;
+            case 'W': write_varname(g, g->V[input[i++] - '0']); continue;
+            case 'L': write_literal_string(g, g->L[input[i++] - '0']); continue;
+            case '+': g->margin++; continue;
+            case '-': g->margin--; continue;
+            case 'n': write_string(g, g->options->name); continue;
+        }
+    }
+}
+
+static void w(struct generator * g, const char * s) {
+    writef(g, s, 0);
+}
+
+static void generate_AE(struct generator * g, struct node * p) {
+    const char * s;
+    switch (p->type) {
+        case c_name:
+            write_varref(g, p->name); break;
+        case c_number:
+            write_int(g, p->number); break;
+        case c_maxint:
+            write_string(g, "9007199254740991"); break;
+        case c_minint:
+            write_string(g, "-9007199254740991"); break;
+        case c_neg:
+            write_char(g, '-'); generate_AE(g, p->right); break;
+        case c_multiply:
+            s = " * "; goto label0;
+        case c_plus:
+            s = " + "; goto label0;
+        case c_minus:
+            s = " - "; goto label0;
+        case c_divide:
+            s = " / ";
```

Comment:
The [snowman](https://snowballstem.org/compiler/snowman.html) doesn't explicitly spell it out, but the type is called `integer`, and snowman does say "/ [...] with the same binding powers and semantics as C" so I think it's clear that it should be integer division.

@dscorbett I guess it's the weakly typed languages which get this wrong?  So Javascript and Python (but not Python 2).

Reply:
Yes, and also Pascal.

Reply:
I've now also addressed this for Javascript and Python, and highlighted this point in the section about implementing a new backend in `CONTRIBUTING.rst`.

It would be good to have a test for this (and indeed other language features, especially those not exercised by any current algorithm implementation) which gets run under each backend, but the current test framework doesn't easily support that.  I'll open a new ticket for that so it doesn't get forgotten.

Reply:
Bringing us back from this side discussion, this should be integer division.